bin/cheribsdtest/cheribsdtest.c

```diff
@@ -431,6 +449,12 @@ cheribsdtest_run_test(const struct cheri_test *ctp)
 		    ccsp->ccs_si_trapno);
 		goto fail;
 	}
+	if ((ctp->ct_flags & CT_FLAG_SI_ADDR) &&
+	    ccsp->ccs_si_addr_expected != ccsp->ccs_si_addr) {
```

Comment:
Do we want exact equality here? Page fault-like things get a null-derived virtual address, and we should probably be checking we haven't somehow got a capability here (it would be quite difficult for CheriBSD to screw that up unless it just gave xEPCC instead, but something like qemu-user could easily make that a mistake).

Reply:
Good question. Does clang use exact equality of capabilities for pointer equality in purecap mode?

Reply:
No (and doing so goes against the spirit of the C spec, since it explicitly calls out the adjacent object case when it comes to pointer equality as the one time where two pointers for different objects can compare equal), you need to use __builtin_cheri_equal_exact. We don't currently have a wrapper in cheric.h as it's only used in one place currently (itself in cheric.h) but you could add one.

Reply:
Some traps do store pc in si_addr rather than a faulting address FWIW.  Note that in a signal handler you can see the saved pc in the ucontext_t with tags, etc. already as well (though si_addr is available to other processes via ptrace()).  I'm not sure we go out of our way to strip tags when storing pc in si_addr.  That said, I'm pretty sure clang doesn't use exact equal but only compares addresses anyway.  I'd almost rather have this case be an exact equals case and force the test to set the expected_si_addr to what it expects to avoid not noticing if we leak a tag we don't expect.